File: sdks/typescript/src/merkle.ts
/** SHA-256 sorted-pair Merkle tree — spec §7. Mirrors protocol/merkle.py:
 * raw 32-byte leaves, ordered-pair concat, odd leaf promoted (not duplicated),
 * direction-free proofs. */

import { sha256 } from "@noble/hashes/sha2.js";
import { ReceiptError } from "./errors.js";

export function sha256Hex(data: Uint8Array): string {
  return "0x" + Buffer.from(sha256(data)).toString("hex");
}

export function hexToBytes(hex: string): Uint8Array {
  if (!/^0x[0-9a-f]{64}$/.test(hex)) throw new ReceiptError("invalid_hash", hex);
  return Uint8Array.from(Buffer.from(hex.slice(2), "hex"));
}

function cmpBytes(a: Uint8Array, b: Uint8Array): number {
  const m = Math.min(a.length, b.length);
  for (let i = 0; i < m; i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}

export function merkleRoot(leaves: Uint8Array[]): string {
  if (leaves.length === 0) throw new ReceiptError("malformed_document", "no leaves");
  let level = [...leaves];
  while (level.length > 1) {
    const next: Uint8Array[] = [];
    for (let i = 0; i < level.length; i += 2) {
      if (i + 1 === level.length) {
        next.push(level[i]); // promote odd leaf
      } else {
        const [a, b] = [level[i], level[i + 1]].sort(cmpBytes);
        next.push(sha256(concat(a, b)));
      }
    }
    level = next;
  }
  return "0x" + Buffer.from(level[0]).toString("hex");
}

function concat(a: Uint8Array, b: Uint8Array): Uint8Array {
  const out = new Uint8Array(a.length + b.length);
  out.set(a, 0);
  out.set(b, a.length);
  return out;
}

/** Direction-free sibling path from leaf index to root (spec §8). */
export function merkleProof(leaves: Uint8Array[], index: number): string[] {
  if (index < 0 || index >= leaves.length) {
    throw new ReceiptError("malformed_proof", "leaf index out of range");
  }
  const siblings: string[] = [];
  let level = [...leaves];
  let idx = index;
  while (level.length > 1) {
    const next: Uint8Array[] = [];
    const sib = idx % 2 === 0 ? idx + 1 : idx - 1;
    for (let i = 0; i < level.length; i += 2) {
      if (i + 1 === level.length) {
        next.push(level[i]);
      } else {
        const [a, b] = [level[i], level[i + 1]].sort(cmpBytes);
        next.push(sha256(concat(a, b)));
      }
    }
    if (sib < level.length) {
      siblings.push("0x" + Buffer.from(level[sib]).toString("hex"));
    }
    idx = Math.floor(idx / 2);
    level = next;
  }
  return siblings;
}

export function verifyMerkleProof(root: string, leaf: Uint8Array, siblings: string[]): boolean {
  let node = leaf;
  for (const sib of siblings) {
    const s = hexToBytes(sib);
    const [a, b] = [node, s].sort(cmpBytes);
    node = sha256(concat(a, b));
  }
  return "0x" + Buffer.from(node).toString("hex") === root;
}

File: sdks/rust/src/merkle.rs
//! SHA-256 sorted-pair Merkle tree — spec §6. Mirrors `protocol/merkle.py`:
//! raw 32-byte leaves, ordered-pair concat, odd node promoted (never
//! duplicated), direction-free proofs.

use crate::errors::{E_BAD_HASH, Result, ReceiptError};
use sha2::{Digest, Sha256};

pub fn sha256_hex(data: &[u8]) -> String {
    format!("0x{}", hex_encode(&Sha256::digest(data)))
}

pub fn hex_encode(b: &[u8]) -> String {
    b.iter().map(|x| format!("{x:02x}")).collect()
}

pub fn hex_decode(s: &str) -> Result<Vec<u8>> {
    let raw = s.strip_prefix("0x").unwrap_or(s);
    if raw.len() % 2 != 0 || !raw.chars().all(|c| c.is_ascii_hexdigit()) {
        return Err(ReceiptError::new(E_BAD_HASH, format!("{s} is not hex")));
    }
    (0..raw.len())
        .step_by(2)
        .map(|i| {
            u8::from_str_radix(&raw[i..i + 2], 16)
                .map_err(|_| ReceiptError::new(E_BAD_HASH, format!("{s} is not hex")))
        })
        .collect()
}

fn parent(a: &[u8; 32], b: &[u8; 32]) -> [u8; 32] {
    let (lo, hi) = if a <= b { (a, b) } else { (b, a) };
    let mut h = Sha256::new();
    h.update(lo);
    h.update(hi);
    h.finalize().into()
}

pub fn merkle_root(leaves: &[[u8; 32]]) -> [u8; 32] {
    assert!(!leaves.is_empty(), "merkle root of empty set");
    let mut level: Vec<[u8; 32]> = leaves.to_vec();
    while level.len() > 1 {
        let mut next = Vec::with_capacity(level.len().div_ceil(2));
        for pair in level.chunks(2) {
            if pair.len() == 1 {
                next.push(pair[0]); // promote odd node
            } else {
                next.push(parent(&pair[0], &pair[1]));
            }
        }
        level = next;
    }
    level[0]
}

/// Direction-free sibling path (bottom-up). Promoted odd nodes contribute
/// no sibling — mirrors Python exactly.
pub fn merkle_proof(leaves: &[[u8; 32]], index: usize) -> Vec<[u8; 32]> {
    assert!(index < leaves.len(), "leaf index out of range");
    let mut siblings = Vec::new();
    let mut level: Vec<[u8; 32]> = leaves.to_vec();
    let mut idx = index;
    while level.len() > 1 {
        let sib = if idx % 2 == 0 { idx + 1 } else { idx - 1 };
        if sib < level.len() {
            siblings.push(level[sib]);
        }
        let mut next = Vec::with_capacity(level.len().div_ceil(2));
        for pair in level.chunks(2) {
            if pair.len() == 1 {
                next.push(pair[0]);
            } else {
                next.push(parent(&pair[0], &pair[1]));
            }
        }
        idx /= 2;
        level = next;
    }
    siblings
}

pub fn verify_proof(leaf: &[u8; 32], siblings: &[[u8; 32]], root: &[u8; 32]) -> bool {
    let mut node = *leaf;
    for sib in siblings {
        node = parent(&node, sib);
    }
    node == *root
}
